src/shared/pkcs11-util.c

```diff
@@ -1232,9 +1401,45 @@ int pkcs11_crypt_device_callback(
          * shouldn't hurt. */
         (void) pkcs11_token_acquire_rng(m, session);
 
-        r = pkcs11_token_find_private_key(m, session, uri, &object);
+        r = pkcs11_token_find_private_key(m, session, uri, &object, &type);
         if (r < 0)
                 return r;
+#if HAVE_OPENSSL
+        if (type == CKK_EC) {
+                _cleanup_(EVP_PKEY_freep) EVP_PKEY *pkey = NULL;
+                _cleanup_(erase_and_freep) uint8_t *shared_secret = NULL;
+                size_t shared_secret_len = 0;
+                d2i_PUBKEY(&pkey, (const uint8_t **)&data->encrypted_key, data->encrypted_key_size);
+                if (!pkey)
+                        return -EIO;
+
+                r = pkcs11_token_ecdh_derive_shared_secret(m, session, object, pkey, NULL, &shared_secret_len);
+                if (r < 0)
+                        return r;
+                if (shared_secret_len == 0)
+                        return -EINVAL;
+
+                shared_secret = malloc(shared_secret_len);
+                if (!shared_secret)
+                        return log_oom();
+
+                r = pkcs11_token_ecdh_derive_shared_secret(m, session, object, pkey, shared_secret, &shared_secret_len);
+                if (r < 0)
+                        return r;
+
+                data->decrypted_key = malloc(72);
+                if (!data->decrypted_key)
+                        return log_oom();
+
+                data->decrypted_key_size = 72;
+                r = hkdf_sha512(shared_secret, shared_secret_len, data->decrypted_key_size, data->decrypted_key);
+                if (r < 0)
+                        return r;
+        }
```

Comment:
this whole if block we already have seen somewhere else ust like this, no? share the code?